Add function to Utility Meter to not reset on a timely basis

gigatexel opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Sometimes, you want your Utility Meter just not to reset automatically.
For instance, if your want a Utility Meter to track your yearly consumption, you need a manual reset of UM - not an automatic one.
Let's say your yearly registration your electricity meter is the beginning of January, one year it will be on the 4th, an other year it will be on the 10th. In this case a YEARLY reset with an offset of a certain amount of days will not work, because it is never fixed.
Thanks to the addition of "NONE" (no reset), a user can decide to use the reset function to manually reset the UM when necessary.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
utility_meter:
  yearly_energy:
    source: sensor.energy
    cycle: none

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/16087

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works. (https://github.com/gigatexel/core/blob/dev/tests/components/utility_meter/test_sensor.py)

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [X] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

gigatexel committed:
Update const.py

gigatexel committed:
Update sensor.py

homeassistant commented:
Hi @gigatexel, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#44760).

Once you do that we will be able to review and accept this pull request.

Thanks!

gigatexel committed:
Update sensor.py

homeassistant commented:
Hi @gigatexel, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#44760).

Once you do that we will be able to review and accept this pull request.

Thanks!

gigatexel committed:
Update sensor.py

homeassistant commented:
Hi @gigatexel, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#44760).

Once you do that we will be able to review and accept this pull request.

Thanks!

gigatexel committed:
Update sensor.py

homeassistant commented:
Hi @gigatexel, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#44760).

Once you do that we will be able to review and accept this pull request.

Thanks!

dgomes reviewed (changes requested):
All good! Just add a test for this new period

gigatexel commented:
> All good! Just add a test for this new period

Any suggestions on how to do that. I'ts my first PR on HA Core. I've checked the docs but I don't find them very clarifying.

Also, who's responsible to update https://www.home-assistant.io/integrations/utility_meter/ ?

dgomes commented:
Have a look into https://github.com/home-assistant/core/blob/dev/tests/components/utility_meter/test_sensor.py 

And you are responsible for updating the documentation (that is also a requirement for this PR to get accepted)

gigatexel committed:
Update test_sensor.py

gigatexel commented:
> Have a look into https://github.com/home-assistant/core/blob/dev/tests/components/utility_meter/test_sensor.py
> 
> And you are responsible for updating the documentation (that is also a requirement for this PR to get accepted)

Test added and docs updated. Linked to both in the topicstart.

dgomes commented:
The test you added does not test the new cycle

You should link the PR to the documentation repository, not your own repository.

gigatexel committed:
Update test_sensor.py

gigatexel committed:
Update test_sensor.py

gigatexel commented:
> The test you added does not test the new cycle
> 
> You should link the PR to the documentation repository, not your own repository.

I added the `https://github.com/home-assistant/core/pull/44760#issuecomment-753472838

> The test you added does not test the new cycle
> 
> You should link the PR to the documentation repository, not your own repository.

Fixed that.
Thanks a lot for your assistance!

gigatexel committed:
Update test_sensor.py

gigatexel committed:
Update test_sensor.py

dgomes commented:
It's not just creating your fork of the documentation... you need to open the PR at https://github.com/home-assistant/home-assistant.io/pulls

gigatexel commented:
> It's not just creating your fork of the documentation... you need to open the PR at https://github.com/home-assistant/home-assistant.io/pulls

Done that and added link to the PR above.

gigatexel committed:
Update test_sensor.py

gigatexel committed:
Update test_sensor.py

gigatexel committed:
Update test_sensor.py

gigatexel committed:
Update test_sensor.py

gigatexel committed:
Update test_sensor.py

gigatexel committed:
Update sensor.py

gigatexel committed:
Update sensor.py

dgomes commented on homeassistant/components/utility_meter/sensor.py:
```diff
@@ -260,7 +261,8 @@ async def async_added_to_hass(self):
                 second=self._period_offset.seconds % 3600 % 60,
             )
-
+        elif self._period == NONE:
```
This is not necessary (it does nothing!)

gigatexel replied:
That's right, but if I omit this the code does not pass, because the const NONE is never used.
When NONE is selected, there is no need to create an `async_track_time_change`, but not using this value at all is a no-pass

dgomes replied:
don't worry, just delete this as it is unnecessary

gigatexel commented:
I did changes in code on the web-UI so I cannot rebase. Setting up VSC and Docker is unfortunately not that very well documented. The HA manuals explain how to set it up, but not how to actually use it.

gigatexel committed:
Update sensor.py

dgomes commented on homeassistant/components/utility_meter/sensor.py:
```diff
@@ -36,6 +36,7 @@
     DATA_UTILITY,
     HOURLY,
     MONTHLY,
+    NONE,
```
unused...

gigatexel replied:
So what do you suggest then? I think there should always be a cycle specified by the user so we need a new constant "NONE". Making the `cycle`optional is not a good idea in my opinion.
But the UM does not need to do anything with it, that's why I added the "pass" in the previous version of the code (also to make sure the tests don't throw a "not used" error.
In the previous revieuw you stated that I should not worry about the tests not passing and so I removed the `pass`section in the code.
Seems like we're running in circles here.

dgomes replied:
in this file you are just importing the variable which is not used...

Do you program in python :) ?

gigatexel replied:
I'm by far not a python expert - I have a few degrees in C++ and Java, but the discussion is not relevant to the coding language.

I think that adding an option to UM that enables the user to have a UM that does not reset automatically has significant value.
There are 3 options:

1. Make `cycle` an optional parameter in the integration
2. Import the parameter and when `async_added_to_hass`runs, make sure no `async_track_time_change` is created (what I did in the previous version by adding `pass`)
3. Import the parameter and do not use it

If you have any better suggestion(s), please let me know - but now I've switched from 2 to 3 based on your feedback, and now 3 is not ok.

dgomes replied:
We are probably getting lost in translation 

I'm not discussing the need for the `"none"` cycle, it has come up in the past in user forums the suggestion to add it and I agree with it.

You only need `NONE` to be defined in the `consts.py` in order for the configuration validation to pass.

BTW: we are very short of implementing option 1. (just a matter of defining `none` as the default cycle) and doing the proper checks
